src/openforms/registrations/contrib/objects_api/utils.py

```diff
@@ -0,0 +1,19 @@
+from django.utils.html import escape
+
+from openforms.typing import JSONValue
+
+
+def escape_html_manually(input: dict) -> dict[str, JSONValue]:
```

Comment:
I don't like that this function has side effects (mutates the dict) *and* also returns values, that makes things confusing. IMO this should be written differently - and it can be really concise:

```py
from typing import TypeVar
from django.utils.html import escape
from openforms.typing import JSONValue

T = TypeVar("T", bound=JSONValue)


def html_escape_json(value: T) -> T:
    """
    Recursively apply HTML escaping to string value nodes.
    """
    match value:
        case list():
            return [html_escape_json(item) for item in value]
        case dict():
            return {key: html_escape_json(value) for key, value in value.items()}
        case str():
            return escape(value)
        case _:
            # nothing to do, return unmodified
            return value
```

The generic type hint declares that whatever input type you give it, you will get back the same return type - and the types are limited by valid JSON-serializable types.

This approach is a pure function without side-effects/mutations :)